Platform/OpenNtfsDxe/Compression.c

```diff
@@ -0,0 +1,685 @@
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  BlockLength = (BlockParameters & BLOCK_LENGTH_BITS) + 1;
+
+  if (Dest != NULL) {
+    if ((BlockParameters & IS_COMPRESSED_BLOCK) != 0) {
+      ClearTextPointer = 0;
+      Tokens           = 0;
+      TagsByte         = 0;
+      while (BlockLength > 0) {
+        if (ClearTextPointer > COMPRESSION_BLOCK) {
+          DEBUG ((DEBUG_INFO, "NTFS: Compression block too large\n"));
+          return EFI_VOLUME_CORRUPTED;
+        }
+
+        if (Tokens == 0) {
+          Status = GetDataRunByte (Clusters, &TagsByte);
+          if (EFI_ERROR (Status)) {
+            return Status;
+          }
+
+          Tokens = 8;
+          --BlockLength;
+          if (BlockLength == 0) {
+            break;
+          }
+        }
+
+        if ((TagsByte & 1U) == 1) {
+          //
+          // Back-reference
+          //
+          Status = GetTwoDataRunBytes (Clusters, &BackReference);
+          if (EFI_ERROR (Status)) {
+            return Status;
+          }
+
+          BlockLength -= 2;
+
+          if (ClearTextPointer == 0) {
+            DEBUG ((DEBUG_INFO, "NTFS: Nontext window empty\n"));
+            return EFI_VOLUME_CORRUPTED;
+          }
+
+          Lmask = BLOCK_LENGTH_BITS;
+          Dshift = 12;
+          for (Index = ClearTextPointer - 1; Index >= 0x10; Index >>= 1) {
+            Lmask >>= 1U;
+            --Dshift;
+          }
+
+          Delta = BackReference >> Dshift;
+          Length = (BackReference & Lmask) + 3;
+
+          if ((Delta > (ClearTextPointer - 1)) || (Length >= COMPRESSION_BLOCK)) {
+            DEBUG ((DEBUG_INFO, "NTFS: Invalid back-reference.\n"));
+            return EFI_VOLUME_CORRUPTED;
+          }
+
+          if (mBufferSize < Length) {
+            DEBUG ((DEBUG_INFO, "NTFS: (DecompressBlock #1) Buffer overflow.\n"));
+            return EFI_VOLUME_CORRUPTED;
+          }
+
+          for (Index = 0; Index < Length; ++Index) {
+            Dest[ClearTextPointer] = Dest[ClearTextPointer - Delta - 1];
+            ++ClearTextPointer;
+          }
+
+          mBufferSize -= Length;
+        } else {
+          //
+          // Plain text
+          //
+          Status = GetDataRunByte (Clusters, &PlainText);
+          if (EFI_ERROR (Status)) {
+            return Status;
+          }
+
+          if (mBufferSize == 0) {
+            DEBUG ((DEBUG_INFO, "NTFS: (DecompressBlock #2) Buffer overflow.\n"));
+            return EFI_VOLUME_CORRUPTED;
+          }
+
+          Dest[ClearTextPointer++] = PlainText;
+
+          --BlockLength;
+          --mBufferSize;
+        }
+
+        TagsByte >>= 1U;
+        --Tokens;
+      }
+
+      return EFI_SUCCESS;
+    }
+
+    if (BlockLength != COMPRESSION_BLOCK) {
+      DEBUG ((DEBUG_INFO, "NTFS: Invalid compression block size %d\n", BlockLength));
+      return EFI_VOLUME_CORRUPTED;
+    }
+  }
+
+  while (BlockLength > 0) {
+    SpareBytes = mClusterSize - Clusters->ClusterOffset;
```

Comment:
This suggests to me there is an invariant that mClusterSize >(=) Clusters->ClusterOffset. Is this correct?

Reply:
That's right.